### PR TITLE
CORTX-30966: Support SSL Certificate as a Secret

### PR DESCRIPTION
**Pull request by walterlopatka**
## Description

This change updates the deploy support for the SSL certficate that is used for cortx-control (CSM) and cortx-server (RGW) pods.

In solution.yaml, if common.ssl.secret is specified, then a pre-built SSL certifiate will be used to create the Secret that holds the certificate.

Alternatively, if common.ssl.externa_secret is specified, then CORTX will use the certificate contained in the specified
secret.

The specified secret must be in the form of a pem file that contains both the certifcate and private key.

Note: It would be better to generate a certificate rather than use a previously-generated one.  This PR does not address this, but rather continues the current process of using the previously-generated certificate.  We can work on that later.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds new functionality)
- [ ] Breaking change (bug fix or new feature that breaks existing functionality)
- [ ] Third-party dependency update
- [ ] Documentation additions or improvements
- [ ] Code quality improvements to existing code or test additions/updates

## Applicable issues

- This change fixes an issue: CORTX-30966

## CORTX image version requirements
N/A

## How was this tested?
- For each of the following scenarios I deployed, created an IAM use (CSM), ran S3 IO, and destroyed the cluster
  - deploy script creates secret:
    - default config: "cortx-cert", "cortx.pem"
    - alternate secret name
    - alternate secret key
  - use external secret
- Tested solution.yaml validation

## Checklist

- [x] The change is tested and works locally.
- [ ] New or changed settings in the solution YAML are documented clearly in the README.md file.
- [x] All commits are signed off and are in agreement with the [CORTX Community DCO and CLA policy](https://github.com/Seagate/cortx/blob/main/doc/dco_cla.md).

If this change requires newer CORTX or third party image versions:

- [ ] The `image` fields in [solution.example.yaml](../k8_cortx_cloud/solution.example.yaml) have been updated to use the required versions.
- [ ] The `appVersion` field of the [Helm chart](../charts/cortx/Chart.yaml) has been updated to use the new CORTX version.

If this change addresses a CORTX Jira issue:

- [x] The title of the PR starts with the issue ID (e.g. `CORTX-XXXXX:`)

### CODE REVIEW

**Comment by walterlopatka** [comment]
I still need to update the documentation.

File: charts/cortx/templates/_config.yaml.tpl [hunk 1]
```diff
@@ -58,7 +58,7 @@ cortx:
       local: /etc/cortx
       config: /etc/cortx
     security:
-      ssl_certificate: /etc/cortx/solution/ssl/s3.seagate.com.pem
```
Comment (walterlopatka) [inline charts/cortx/templates/_config.yaml.tpl:61]: This name could be anything.  "s3.seagate.com.pem" didn't seem right, so I changed it to a simple "cortx.pem".

**Review by osowski — COMMENTED** [review]
Overall it looks good. Just a question on the parameters getting added. I would expect it to be the case of:
`common.ssl.secret` || (`common.ssl.external_secret` && `common.ssl.secret_key`)

I don't think it makes sense from a user experience to require the user the specify what the name of the key is inside of the Secret we are going to generate. It would make more sense for us to just pick one we'd prefer to use already.

@keithpine thoughts?

File: k8_cortx_cloud/deploy-cortx-cloud.sh [hunk 2]
```diff
@@ -612,6 +625,22 @@ function deployCortxSecrets()
...
+    # Create cortx-cert Secret
+    cortx_secret_cert_name=$(getSolutionValue "solution.common.ssl.secret")  # This is a global variable
+    cortx_secret_cert_key=$(getSolutionValue "solution.common.ssl.secret_key")  # This is a global variable
+    if [[ -n "${cortx_secret_cert_name}" ]]; then
+        if ! kubectl create secret generic "${cortx_secret_cert_name}" \
+            --from-file="${cortx_secret_cert_key}=./ssl-cert/cortx.pem" \
+            --namespace="${namespace}"; then
+            printf "Exit early.  Failed to create Secret '%s'\n" "${cortx_secret_cert_name}"
+            exit 1
+        fi
```
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:638]: Don't need this if using a Helm-managed secret.

File: README.md [hunk 3]
```diff
@@ -314,6 +314,9 @@ This section contains common parameters that affect all CORTX components running
 | Name                                                  | Description                                                     | Default Value       |
 | ----------------------------------------------------- | --------------------------------------------------------------- | ------------------- |
 | `common.storage_provisioner_path`                     | TODO       | `/mnt/fs-local-volume` |
+| `common.ssl.secret`                                   | Name of a Secret to be generated by `deploy-cortx-cloud.sh` that contains a self-signed certificate.<br />Note: Either `common.ssl.secret` or `common.ssl.external_secret` must be specified. | `""` |
```
Comment (keithpine) [inline README.md:317]: I don't think this option is necessary. Why don't we just use the name we already do, which is derived from the release ({release name}-ssl-cert)?

File: README.md [hunk 4]
```diff
@@ -314,6 +314,9 @@ This section contains common parameters that affect all CORTX components running
...
+| `common.ssl.external_secret`                          | Names of an existing Secret that contains a TLS/SSL certificate.  This certificate must be in PEM format and must contain the Private Key and Certificate.<br />Note: Either `common.ssl.secret` or `common.ssl.external_secret` must be specified. | `""` |
```
Comment (keithpine) [inline README.md:318]:
```suggestion
| `common.ssl.external_secret`                          | Name of an existing Secret that contains a TLS/SSL certificate. The certificate is a single file in PEM form that includes the full certificate chain and private key in PEM format. If the external secret is not provided, a self-signed certificate will be used instead. | `""` |
```

File: charts/cortx/templates/_config.yaml.tpl [hunk 5]
```diff
@@ -58,7 +58,7 @@ cortx:
...
+      ssl_certificate: /etc/cortx/solution/ssl/cortx.pem
       domain_certificate: /etc/cortx/solution/ssl/stx.pem
       device_certificate: /etc/cortx/solution/ssl/stx.pem
```
Comment (keithpine) [inline charts/cortx/templates/_config.yaml.tpl:63]: Unrelated, but are these necessary?

File: charts/cortx/templates/control/deployment.yaml [hunk 6]
```diff
@@ -25,8 +25,11 @@ spec:
           configMap:
             name: {{ include "cortx.configmapName" . }}
         - name: cortx-ssl-cert
-          configMap:
-            name: {{ include "cortx.tls.configmapName" . }}
+          secret:
+            secretName: {{ .Values.control.certificateSecret }}
```
Comment (keithpine) [inline charts/cortx/templates/control/deployment.yaml:29]: Instead, rename template `cortx.tls.configmapName` to `cortx.tls.secretName`, and derive it based on either the release name, or use the externally provided name.

File: charts/cortx/README.md [hunk 7]
```diff
@@ -154,6 +156,8 @@ helm uninstall cortx
 | rbac.create | bool | `true` | Enable the creation of RBAC resources, Role and RoleBinding, for the CORTX ServiceAccount. |
 | server.auth.adminAccessKey | string | `"cortx-admin"` | The admin user's Access Key |
 | server.auth.adminUser | string | `"cortx-user"` | Name of the admin user that is created when initializing the cluster |
+| server.certificateSecret | string | `""` | Name of Secret that contains the certificate |
+| server.certificateSecretKey | string | `""` | Name of Secret key that contains the certificate |
```
Comment (keithpine) [inline charts/cortx/README.md:160]: Is it really necessary to provide this flexibility?

Say one day we can upgrade to using a real tlsSecret, those fields are fixed anyways. Why not just require a fixed key name, and just document that.

File: charts/cortx/values.yaml [hunk 8]
```diff
@@ -219,6 +219,10 @@ control:
   enabled: true
   # -- Number of Control replicas
   replicaCount: 1
+  # -- Name of Secret that contains the certificate
+  certificateSecret: ""
```
Comment (keithpine) [inline charts/cortx/values.yaml:223]: Remove in favor of using auto-generated secret name.

File: charts/cortx/values.yaml [hunk 9]
```diff
@@ -352,6 +356,10 @@ server:
   enabled: true
   # -- Number of Server replicas
   replicaCount: 1
+  # -- Name of Secret that contains the certificate
+  certificateSecret: ""
```
Comment (keithpine) [inline charts/cortx/values.yaml:360]: Remove in favor of using auto-generated secret name.

File: charts/cortx/values.yaml [hunk 10]
```diff
@@ -219,6 +219,10 @@ control:
   enabled: true
   # -- Number of Control replicas
   replicaCount: 1
+  # -- Name of Secret that contains the certificate
+  certificateSecret: ""
+  # -- Name of Secret key that contains the certificate
+  certificateSecretKey: ""
```
Comment (keithpine) [inline charts/cortx/values.yaml:225]: I like the idea of being able to specify these separately, but I don't think CORTX allows this? The cluster configuration only provides a single certificate value for all components. What would happen if you actually used two different secrets?

I think this has to be a single value at the top-level.

File: k8_cortx_cloud/deploy-cortx-cloud.sh [hunk 11]
```diff
@@ -142,6 +142,17 @@ function configurationCheck()
...
+    secret_name=$(getSolutionValue "solution.common.ssl.secret")
+    secret_ext=$(getSolutionValue "solution.common.ssl.external_secret")
+    if [[ -z "${secret_name}" && -z "${secret_ext}" ]] ; then
+        printf "Error: %s: solution.common.ssl.secret or solution.common.ssl.external_secret must be specified\n" "${solution_yaml}"
+        exit 1
+    elif [[ -n "${secret_name}" && -n "${secret_ext}" ]] ; then
+        printf "Error: %s: Cannot specify both solution.common.ssl.secret or solution.common.ssl.external_secret\n" "${solution_yaml}"
+        exit 1
+    fi
+
```
Comment (keithpine) [inline k8_cortx_cloud/deploy-cortx-cloud.sh:155]: Don't need this if using a Helm-managed secret.

File: k8_cortx_cloud/destroy-cortx-cloud.sh [hunk 12]
```diff
@@ -214,13 +214,21 @@ function deleteSecrets()
...
 
         [[ -d $(pwd)/cortx-cloud-helm-pkg ]] && find "$(pwd)/cortx-cloud-helm-pkg" -name "secret-info.txt" -delete
     fi
+
+    local ssl_secret_name
+    ssl_secret_name=$(./parse_scripts/parse_yaml.sh "${solution_yaml}" "solution.common.ssl.secret")
+    if [[ -n "${ssl_secret_name}" ]]; then
+        ssl_secret_name=$(echo "${ssl_secret_name}" | cut -f2 -d'>')
+        kubectl delete secret "${ssl_secret_name}" --namespace="${namespace}" --ignore-not-found=true
+    fi
```
Comment (keithpine) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:231]: Don't need this if using a Helm-managed secret.

File: k8_cortx_cloud/solution.example.yaml [hunk 13]
```diff
@@ -23,6 +23,9 @@ solution:
     busybox: ghcr.io/seagate/busybox:latest
   common:
     storage_provisioner_path: /mnt/fs-local-volume
+    ssl:
+       secret: cortx-cert
+       secret_key: cortx.pem
```
Comment (keithpine) [inline k8_cortx_cloud/solution.example.yaml:28]: Don't need this if using a Helm-managed secret.

File: k8_cortx_cloud/solution_validation_scripts/solution-check.yaml [hunk 14]
```diff
@@ -12,6 +12,8 @@ solution:
     rancher: required
   common:
     storage_provisioner_path: required
+    ssl:
+      secret_key: required
```
Comment (keithpine) [inline k8_cortx_cloud/solution_validation_scripts/solution-check.yaml:16]: Don't need this if using a Helm-managed secret.

**Review by keithpine — CHANGES_REQUESTED** [review]
I'd prefer to see the Secret continue to be managed by Helm. If that's not feasible, let me know.

**Comment by walterlopatka** [comment]
Closing at this PR is hopeless.  I have resubmitted #362

**Pull request closed without merging**